router/throttler/throttler.go

```diff
@@ -1,173 +1,224 @@
 package throttler
 
 import (
+	"context"
 	"fmt"
+	"sync"
 	"time"
 
+	"github.com/go-redis/redis/v8"
+
 	"github.com/rudderlabs/rudder-server/config"
-	"github.com/rudderlabs/rudder-server/router/throttler/ratelimiter"
+	"github.com/rudderlabs/rudder-server/internal/throttling"
 	"github.com/rudderlabs/rudder-server/utils/logger"
 )
 
 const (
-	DESTINATION_LEVEL = "destination"
-	USER_LEVEL        = "user"
-	ALL_LEVELS        = "all"
+	algoTypeGoRate         = "gorate"
+	algoTypeGCRA           = "gcra"
+	algoTypeRedisGCRA      = "redis-gcra"
+	algoTypeRedisSortedSet = "redis-sorted-set"
 )
 
-// Throttler is an interface for throttling functions
-type Throttler interface {
-	CheckLimitReached(destID, userID string, currentTime time.Time) bool
-	Inc(destID, userID string, currentTime time.Time)
-	Dec(destID, userID string, count int64, currentTime time.Time, atLevel string)
-	IsEnabled() bool
-	IsUserLevelEnabled() bool
-	IsDestLevelEnabled() bool
+var clientsPool redisClientsPool
+
+type limiter interface {
+	// Limit should return true if the request can be done or false if the limit is reached.
+	Limit(ctx context.Context, cost, rate, window int64, key string) (
+		allowed bool, ret throttling.TokenReturner, err error,
+	)
 }
 
-type Limiter struct {
+type limiterSettings struct {
 	enabled     bool
-	eventLimit  int
+	eventLimit  int64
 	timeWindow  time.Duration
-	ratelimiter *ratelimiter.RateLimiter
+	rateLimiter limiter
```

Comment:
`rateLimiter` could be reused across all destinationIDs of the same router. Each destination has its own key, however we seem to be using a separate limiter per key, whereas our limiters support multiple keys.